tests/functional/test_deploy.py

```diff
@@ -0,0 +1,308 @@
+from typing import Generator, Optional, Dict, List, Union
+from functools import lru_cache
+import subprocess
+import tempfile
+import warnings
+import textwrap
+import os.path
+import signal
+import shutil
+import glob
+import time
+import json
+import stat
+import os
+
+
+CWD = os.path.dirname(os.path.abspath(__file__))
+
+
+@lru_cache()
+def get_container_image() -> str:
+    image_name: str = json.loads(
+        subprocess.check_output(
+            [
+                "nix-instantiate",
+                "--strict",
+                os.path.join(CWD, "container"),
+                "--eval",
+                "--json",
+                "-A",
+                "image",
+            ]
+        )
+    )
+
+    image_id = (
+        subprocess.check_output(["docker", "images", "-q", image_name]).decode().strip()
+    )
+    if image_id:
+        return image_id
+
+    warnings.warn("Building NixOS container, this may take some time.")
+
+    store_path: str = subprocess.run(
+        [
+            "nix-build",
+            "--show-trace",
+            "--no-out-link",
+            os.path.join(CWD, "container"),
+            "-A",
+            "tarball",
+        ],
+        check=True,
+        stdout=subprocess.PIPE,
+    ).stdout.decode().strip()
+
+    image_file: str = glob.glob(
+        os.path.join(store_path, "tarball/nixos-system-*.tar.xz")
+    )[0]
+
+    return (
+        subprocess.check_output(["docker", "import", image_file, image_name])
+        .decode()
+        .strip()
+    )
+
+
+class Container:
+
+    name: str
+    image_id: str
+    container_id: Optional[str]
+    ssh_port: int
+    hostname: str
+    env: Dict[str, str]
+
+    def __init__(self, name: str, ssh_port: int, image_id: str):
+        self.name = name
+        self.container_id = None
+        self.image_id = image_id
+        self._process = None
+        self.ssh_port = ssh_port
+        self.hostname = "127.0.0.1"
+        self.user = "root"
+        self.env = {}
+
+    @property
+    def started(self) -> bool:
+        return self.container_id is not None
+
+    def run(self):
+        process = subprocess.run(
+            [
+                "docker",
+                "run",
+                "--privileged",
```

Comment:
So.

Funny thing about this.

`docker` from `podman` makes `--privileged` not so interesting.

`docker` from `docker` makes `--privileged` .......much more interesting.

![signal-2020-04-29-170151](https://user-images.githubusercontent.com/76716/80646681-2f549f80-8a3b-11ea-9de7-7c099a455d94.jpeg)

It actually took over my entire system. There are even 2 gettys running: ee the alternating hostnames.

lol

so I think actually we should figure out how to force this to only use podman, and force it to only use podman in user-mode?

Reply:
We're now only calling `podman`. The CI still runs on Docker because we never got rootless Podman to work on Github Actions.

It does so by putting a symlink from `docker` -> `podman` which is a total hack but OK if only in CI.